### ISSUE TITLE
TLS configuration options

### ISSUE THREAD

**Issue by kurtkincaid** [arangodb/arangodb#1771]
What would be involved in exposing more TLS configuration options? At work, I am deeply involved in encryption every day, so I am probably more concerned about it than some people, but as I see it, there are several options that need to be available.

- Support for TLSv1.2. From a strict security perspective, anything less TLSv1.2 should not be used. It is fine to leave support for lower protocol versions, but support for TLSv1.2 is an absolute requirement. Anything other than GCM cipher suites in TLSv1.2 is vulnerable to one of the POODLE attack variants.
- There should be the ability to specify the Diffie Hellman parameters. This would give the ability to use unique parameters in each ArangoDB instance, rather than relying on the default DH parameters everywhere. (See https://weakdh.org) 

Other options that would be very useful (not as significant as the points above):

- Support for mutual TLS authentication between client and server. This is not intended to be a replacement for username/password authentication, but to restrict access to accredited and certified clients.
- PKCS#11 integration. This would allow storage of server keys on HSMs, Smart Cards, etc. For example, I have ArangoDB running on two servers that are HSM clients. All of the server's other keys (for the web server, for signatures, etc.) are protected in my HSM. The only keys that are not HSM protected are my ArangoDB keys.
- Support for reading server keys out of PKCS#12 files. This should be in addition to the previous point, as HSMs and Smart Cards are not readily available in all cases. A PKCS#12 file (also know as PFX, especially in Windows) is encrypted by definition. True, the passphrase would need to be included in the arrangod.conf file, but this is still preferable to leaving the private key in an unencrypted PEM file.

**Comment by dothebart** [comment]
So as 1) was easy to implement, did that now and backported it to 2.8 and it will be part of 2.8.5. More to come .

**Comment by kurtkincaid** [comment]
Outstanding! Works like a charm.

**Comment by razvanphp** [comment]
Any reason why this is [not documented](https://docs.arangodb.com/ConfigureArango/Arangod.html#ssl-protocol) and why it is not default now?

**Comment by dothebart** [comment]
You're right, the documentation is missing. However depending on your host OS your openssl may not have TLSv1.2 which is why we can't make it default. But we should do this once we drop support for these.

**Comment by kurtkincaid** [comment]
[arangodb_ciphersuites.txt](https://github.com/arangodb/arangodb/files/201598/arangodb_ciphersuites.txt)
[openssl_ciphersuites](https://cloud.githubusercontent.com/assets/219304/14235517/39abc3c8-f9c6-11e5-8d04-ecab96a3240e.png)

I have attached two files. The text file is part of the output generated with [sslyze](https://github.com/nabla-c0d3/sslyze) when pointing at ArangoDB with TLSv1.2 enabled. I have intentionally not specified any cipher suites in the config file. Is there any way to enabled ECDHE cipher suites? They are the only cipher suites that provide Forward Secrecy. (Technically, the DHE cipher suites provide this as well, but they are far too slow to be used in production.)

The PNG file shows a selection of the OpenSSL TLSv1.2 cipher suites. Ideally, only the GCM cipher suites (only available in TLSv1.2) should be used. The cipher suites highlighted in blue are the ones that are currently available in ArangoDB. What would be required to add the cipher suites highlighted in red? It would seem that the current version would preclude the use of elliptic curve certificates for the ArangoDB server, given that RSA is the only supported key exchange option.
For what it is worth, here is the command I ran with sslyze:

`sslyze --certinfo=basic --sslv2 --sslv3 --tlsv1 --tlsv1_1 --tlsv1_2 --hide_rejected_ciphers localhost:8530`

I have tried to explicitly add the cipher suites highlighted in red, and while the server starts just fine, I get a handshake failure when I try to connect.

Thanks for looking into this.

Kurt

**Comment by dothebart** [comment]
hm, to be honest, I'm failing to generate a server.pem that allows TLSv1.2 in first place ;-(

So  sslyze just tells me no cyphers are available at all...

If you add `--server.ssl-cipher-list=ECDHE-RSA-AES256-GCM-SHA384` does that work for you?

**Comment by kurtkincaid** [comment]
No luck. It is interesting that the server starts, but I immediately get a handshake failure during the ServerHello. Here is the error message I get when I try to connect using OpenSSL's s_client command:

> CONNECTED(00000194)
> 18720:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:.\ssl\s23_clnt.c:769:
> 
> no peer certificate available

The server apparently recognizes ECDHE-RSA-AES256-GCM-SHA384 as a cipher suite, though, because when I tried to start it with `--server.ssl-cipher-list=ZZZZZ`, it promptly failed with:

> 2016-04-04T15:03:16Z [17520] ERROR SSL error: error:1410D0B9:SSL routines:SSL_CTX_set_cipher_list:no cipher match
> 2016-04-04T15:03:16Z [17520] FATAL cannot set SSL cipher list 'ZZZZZ'

**Comment by dothebart** [comment]
So I tried to simply specify anything from 

    openssl ciphers -VTLSv1.2:IDHE:INULL:IADH:IDH

in one big list:

    --server.ssl-cipher-list='ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-SHA384,ECDHE-ECDSA-AES256-SHA384,DH-DSS-AES256-GCM-SHA384,DHE-DSS-AES256-GCM-SHA384,DH-RSA-AES256-GCM-SHA384,DHE-RSA-AES256-GCM-SHA384,DHE-RSA-AES256-SHA256,DHE-DSS-AES256-SHA256,DH-RSA-AES256-SHA256,DH-DSS-AES256-SHA256,ADH-AES256-GCM-SHA384,ADH-AES256-SHA256,ECDH-RSA-AES256-GCM-SHA384,ECDH-ECDSA-AES256-GCM-SHA384,ECDH-RSA-AES256-SHA384,ECDH-ECDSA-AES256-SHA384,AES256-GCM-SHA384,AES256-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-SHA256,ECDHE-ECDSA-AES128-SHA256,DH-DSS-AES128-GCM-SHA256,DHE-DSS-AES128-GCM-SHA256,DH-RSA-AES128-GCM-SHA256,DHE-RSA-AES128-GCM-SHA256,DHE-RSA-AES128-SHA256,DHE-DSS-AES128-SHA256,DH-RSA-AES128-SHA256,DH-DSS-AES128-SHA256,ADH-AES128-GCM-SHA256,ADH-AES128-SHA256,ECDH-RSA-AES128-GCM-SHA256,ECDH-ECDSA-AES128-GCM-SHA256,ECDH-RSA-AES128-SHA256,ECDH-ECDSA-AES128-SHA256,AES128-GCM-SHA256,AES128-SHA256'  

Now I'm getting the RSA ones:

  * TLSV1_2 Cipher Suites:
      Preferred:                       
        TLS_RSA_WITH_AES_256_GCM_SHA384                   -              256 bits                                                                  
      Accepted:                        
        TLS_RSA_WITH_AES_256_GCM_SHA384                   -              256 bits                                                                  
        TLS_RSA_WITH_AES_256_CBC_SHA256                   -              256 bits                                                                  
        TLS_RSA_WITH_AES_128_CBC_SHA256                   -              128 bits                                                                  
        TLS_RSA_WITH_AES_128_GCM_SHA256                   -              128 bits

**Comment by dothebart** [comment]
ok, it seems as if one needs to initialize the curves: https://github.com/nginx/nginx/commit/f39642a90798af3c61de6e74b8da6cad96a9f159

**Comment by razvanphp** [comment]
Just to mention here for the reference, the current request module in Foxx also does not support TLS 1.2 (for servers accepting only this).

The error looks like this:
```
ERROR Cause: InternalServerError: Could not connect to 'ssl://server.domain.com:443' 'SSL: during SSL_connect: 104 - Connection reset by peer'
2016-05-13T13:22:13 [787] ERROR     at Module.load (/usr/share/arangodb/js/common/bootstrap/modules.js:527:21)
```

I would also suggest you implement a "strongest first Cipher default algorithm" for the server, so that OSs that support TLS 1.2 get this as default, and older supported openssl versions should gracefully downgrade.

**Comment by baslr** [comment]
#1974 
ArangoDB v3.1.18 fixes a generic TLS related bug.

**Comment by maja42** [comment]
Is it possible to enable mutual TLS authentication? The docs in "ssl.cafile" state the following:

> If the file is specified, The server will only accept client requests with certificates issued by these CAs. Do not specify this option if you want clients to be able to connect without specific certificates.

However, if I configure this option, arangodb still accepts clients that are signed by a different CA.

**Comment by innwerkler** [comment]
Speaking of mutual TLS: If would also be quite helpful to have support for mutual TLS in the Foxx request module, to make requests to services requiring a client certificate.
Unless this is already supported and I just did not figure out how to do it?
